opm/autodiff/AquiferInterface.hpp

```diff
@@ -0,0 +1,224 @@
+      Qai_[idx]/context.dofVolume(spaceIdx, timeIdx);
+    }
+
+    inline Scalar gravity_() const
+    {
+      return ebos_simulator_.problem().gravity()[2];
+    }
+
+    inline void initQuantities(const Aquancon::AquanconOutput& connection)
+    {
+      // We reset the cumulative flux at the start of any simulation, so, W_flux = 0
+      W_flux_ = 0.;
+
+      // We next get our connections to the aquifer and initialize these quantities using the initialize_connections function
+      initializeConnections(connection);
+
+      calculateAquiferCondition();
+
+      pressure_previous_.resize(cell_idx_.size(), 0.);
+      pressure_current_.resize(cell_idx_.size(), 0.);
+      Qai_.resize(cell_idx_.size(), 0.0);
+    }
+
+    inline void updateCellPressure(std::vector<Eval>& pressure_water, const int idx, const IntensiveQuantities& intQuants)
+    {
+      const auto& fs = intQuants.fluidState();
+      pressure_water.at(idx) = fs.pressure(waterPhaseIdx);
+    }
+
+    inline void updateCellPressure(std::vector<Scalar>& pressure_water, const int idx, const IntensiveQuantities& intQuants)
+    {
+      const auto& fs = intQuants.fluidState();
+      pressure_water.at(idx) = fs.pressure(waterPhaseIdx).value();
+    }
+
+    inline void updateCellDensity(const int idx, const IntensiveQuantities& intQuants)
+    {
+      const auto& fs = intQuants.fluidState();
+      rhow_.at(idx) = fs.density(waterPhaseIdx);
+    }
+
+    template<class faceCellType, class ugridType>
+    inline const double getFaceArea(const faceCellType& faceCells, const ugridType& ugrid,
+                                    const int faceIdx, const int idx,
+                                    const Aquancon::AquanconOutput& connection) const
+      {
+        // Check now if the face is outside of the reservoir, or if it adjoins an inactive cell
+        // Do not make the connection if the product of the two cellIdx > 0. This is because the
+        // face is within the reservoir/not connected to boundary. (We still have yet to check for inactive cell adjoining)
+        double faceArea = 0.;
+        const auto cellNeighbour0 = faceCells(faceIdx,0);
+        const auto cellNeighbour1 = faceCells(faceIdx,1);
+        const auto defaultFaceArea = Opm::UgGridHelpers::faceArea(ugrid, faceIdx);
+        const auto calculatedFaceArea = (!connection.influx_coeff.at(idx))?
+        defaultFaceArea :
+        *(connection.influx_coeff.at(idx));
+        faceArea = (cellNeighbour0 * cellNeighbour1 > 0)? 0. : calculatedFaceArea;
+        if (cellNeighbour1 == 0){
+          faceArea = (cellNeighbour0 < 0)? faceArea : 0.;
+        }
+        else if (cellNeighbour0 == 0){
+          faceArea = (cellNeighbour1 < 0)? faceArea : 0.;
+        }
+        return faceArea;
+      }
+
+      virtual void endTimeStep() = 0;
+
+  protected:
+    const Simulator& ebos_simulator_;
+    const Aquancon::AquanconOutput& connection_;

```

Comment:
This can not be reference here.

Reply:
Please fix this reference problem (which will cause `std::bad_alloc` error. ) and also the merging conflicts, so we perform `Jenkins` test to help the review of the PR.

Reply:
I fixed the referencing. But the const Simulator& ebos_simulator remains, else it will give many errors. The problem however is able to run the test cases. Howver, the WReorder mistakes that are fixed in #1714 are back again, and I don't seem to be able to remove them. I have tried many different orders, however they keep popping up.

Reply:
just keep the same order in the initializer list, as the declaration order in the class.

```
int foo;
int bar;

->
FooBar() : foo(1), bar(1) {}
```

Reply:
> const Simulator& ebos_simulator remains,

It should remain as reference. Anyway, it is a very huge class and we do not want to make a copy of it. And it lives much longer than the Aquifer models, so we do not worry it is gone when we use it.